plugin.video.massengeschmack/resources/lib/listing.py

```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
+# 
+# Massengeschmack Kodi add-on
+# Copyright (C) 2013-2016 by Janek Bevendorff
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import xbmcplugin
+import xbmcgui
+
+from globalvars import *
+
+
+class Listing(object):
+    def __init__(self):
+        self.__source = None
+
+    def generate(self, source):
+        """
+        Generate listing from data source.
+        Will return False if the given DataSource does not intend to return a non-empty listing generator,
+        i.e. when L{resources.lib.datasource.DataSource.hasListItems} returns False
+        
+        @type source: resources.lib.datasource.DataSource
+        @param source: the data source object
+        @rtype: bool
+        @return Whether the given DataSource actually intends to return a non-empty listing
+        """
+        if not source.hasListItems():
+            return False
+
+        self.__source = source
+        items         = source.getListItems()
+
+        for i in items:
+            self.__addDir(i)
+
+        return True
+    
+    def show(self):
+        """
+        Show the listing after it has been generated.
+        You should check the output of L{generate} before trying to show a listing.
+        """
+        if 'true' == ADDON.getSetting('advanced.viewmodeFix'):
+            self.setViewMode(510)
+        else:
+            self.setViewMode(55)
+        xbmcplugin.endOfDirectory(ADDON_HANDLE)
+    
+    def setViewMode(self, id):
+        """
+        Set the view mode of the current listing.
+        
+        @type id: int
+        @param id: the view mode ID from the current skin
+        """
+        xbmcplugin.setContent(ADDON_HANDLE, self.__source.getContentMode())
+        xbmc.executebuiltin('Container.SetViewMode(' + str(id) + ')')
+    
+    def __addDir(self, listItem):
+        xbmcListItem = xbmcgui.ListItem(listItem.getData('name'), iconImage=listItem.getData('thumbnail'),
```

Comment:
Also please set the `mediatype` property for all listitems that provide one of these: `"video", "movie", "tvshow", "season", "episode" or "musicvideo"`. See setInfo documentation:

https://codedocs.xyz/AlwinEsch/kodi/group__python__xbmcgui__listitem.html#ga0b71166869bda87ad744942888fb5f14